model/integration_action.go

```diff
@@ -0,0 +1,260 @@
+	PostId     string          `json:"post_id"`
+	TriggerId  string          `json:"trigger_id"`
+	Type       string          `json:"type"`
+	DataSource string          `json:"data_source"`
+	Context    StringInterface `json:"context,omitempty"`
+}
+
+type PostActionIntegrationResponse struct {
+	Update        *Post  `json:"update"`
+	EphemeralText string `json:"ephemeral_text"`
+}
+
+type PostActionAPIResponse struct {
+	Status    string `json:"status"` // needed to maintain backwards compatibility
+	TriggerId string `json:"trigger_id"`
+}
+
+type Dialog struct {
+	CallbackId     string          `json:"callback_id"`
+	Title          string          `json:"title"`
+	Elements       []DialogElement `json:"elements"`
+	SubmitLabel    string          `json:"submit_label"`
+	NotifyOnCancel bool            `json:"notify_on_cancel"`
+	State          string          `json:"state"`
+}
+
+type DialogElement struct {
+	DisplayName string               `json:"display_name"`
+	Name        string               `json:"name"`
+	Type        string               `json:"type"`
+	SubType     string               `json:"subtype"`
+	Default     string               `json:"default"`
+	Placeholder string               `json:"placeholder"`
+	HelpText    string               `json:"help_text"`
+	Optional    bool                 `json:"optional"`
+	MinLength   int                  `json:"min_length"`
+	MaxLength   int                  `json:"max_length"`
+	DataSource  string               `json:"data_source"`
+	Options     []*PostActionOptions `json:"options"`
+}
+
+type OpenDialogRequest struct {
+	TriggerId string `json:"trigger_id"`
+	URL       string `json:"url"`
+	Dialog    Dialog `json:"dialog"`
+}
+
+type SubmitDialogRequest struct {
+	Type       string                 `json:"type"`
+	URL        string                 `json:"url,omitempty"`
+	CallbackId string                 `json:"callback_id"`
+	State      string                 `json:"state"`
+	UserId     string                 `json:"user_id"`
+	ChannelId  string                 `json:"channel_id"`
+	TeamId     string                 `json:"team_id"`
+	Submission map[string]interface{} `json:"submission"`
+}
+
+func (r *PostActionIntegrationRequest) ToJson() string {
+	b, _ := json.Marshal(r)
+	return string(b)
+}
+
+func GenerateTriggerId(userId string, s crypto.Signer) (string, string, *AppError) {
+	clientTriggerId := NewId()
+	triggerData := strings.Join([]string{clientTriggerId, userId, strconv.FormatInt(GetMillis(), 10)}, ":") + ":"
+
+	h := crypto.SHA256
+	sum := h.New()
+	sum.Write([]byte(triggerData))
+	signature, err := s.Sign(rand.Reader, sum.Sum(nil), h)
+	if err != nil {
+		return "", "", NewAppError("GenerateTriggerId", "interactive_message.generate_trigger_id.signing_failed", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	base64Sig := base64.StdEncoding.EncodeToString(signature)
+
+	triggerId := base64.StdEncoding.EncodeToString([]byte(triggerData + base64Sig))
```

Comment:
If I'm parsing this right, it seems like we effectively `base64(base64(signature))`, making the trigger longer than necessary. I wonder if we could `bytes.Join` (and `bytes.Split` on decode) and just encode to base64 once for compactness? It's not a blocker, since this encoding is effectively throwaway -- we can easily change it in the future -- but a simpler encoding might be useful.

Reply:
I actually was originally doing something similar to what you suggest on my first iteration of this code but I was running into issues where sometimes the signature would contain a byte or character that would affect the non-signature part of the trigger ID. Not exactly sure what the issue was but I did this as a way to be sure that wouldn't happen